### PR TITLE
Redo SVD interface and replace implementation with lapack-based imple…

### PR DESCRIPTION
**Pull request by btracey**
…mentation

### CODE REVIEW

**Comment by btracey** [comment]
Needs lapack/85

**Comment by btracey** [comment]
Fixes https://github.com/gonum/matrix/issues/244

**Comment by btracey** [comment]
Fixes https://github.com/gonum/matrix/issues/269

File: mat64/svd.go [hunk 1]
```diff
@@ -6,474 +6,182 @@
...
 )
 
-type SVDFactors struct {
-	U     *Dense
-	Sigma []float64
-	V     *Dense
-	m, n  int
+// SVDKind specifies the treatment of singular vectors during the
+// factorization.
+// NOTE: Should these be in matrix?
```
Comment (kortschak) [inline mat64/svd.go]: Probably. There may be a mat32 (generated), but is a cmat128 SVD likely?

File: mat64/svd.go [hunk 2]
```diff
@@ -6,474 +6,182 @@
...
+//  A = U * Σ * V^T
+// where Σ is an m×n diagonal matrix of singular vectors, U is an m×m unitary
+// matrix of left singular vectors, and V is an n×n matrix of right singular vectors.
 //
-//  sigma[0] >= sigma[1] >= ... >= sigma[n-1].
+// It is frequently not necessary to compute the full SVD. Computation time and
+// storage costs can be reduced using the appropriate kind. Only the singular
+// values can be computed (kind == SVDNone), or a "thin" representation of the
+// singular values (kind = SVDThin). The thin representation can save a significant
+// amount of memory if m >> n. See the documentation for more information.
```
Comment (kortschak) [inline mat64/svd.go]: See what documentation?

File: mat64/svd.go [hunk 3]
```diff
@@ -6,474 +6,182 @@
...
 //
-// The matrix condition number and the effective numerical rank can be computed from
-// this decomposition.
-func SVD(a *Dense, epsilon, small float64, wantu, wantv bool) SVDFactors {
+// Factorize returns whether the decomposition succeeded. If the decomposition
+// failed, routines that require a successful factorization are not available
```
Comment (kortschak) [inline mat64/svd.go]: s/not available/will panic/

File: mat64/svd.go [hunk 4]
```diff
@@ -6,474 +6,182 @@
...
+// (Values, UFromSVD, etc.).
```
Comment (kortschak) [inline mat64/svd.go]: Don't list them here, but do state that Factorise must have been successful or the method will panic at each method.

File: mat64/svd.go [hunk 5]
```diff
@@ -6,474 +6,182 @@
...
-// Rank returns the number of non-negligible singular values in the sigma held by
-// the factorisation with the given epsilon.
-func (f SVDFactors) Rank(epsilon float64) int {
-	if len(f.Sigma) == 0 {
-		return 0
+// UFromSVD extracts the matrix U from the singular value decomposition, storing
+// the result in-place into the receiver. U is size m×m if svd.Kind() == SVDFull,
+// of size m×min(m,n) if svd.Kind() == SVDThin, and UFromSVD panics otherwise.
+func (m *Dense) UFromSVD(svd *SVD) {
+	r := svd.u.Rows
```
Comment (kortschak) [inline mat64/svd.go]: Early panic for kind == 0.

File: mat64/svd.go [hunk 6]
```diff
@@ -6,474 +6,182 @@
...
 }
 
-// Cond returns the 2-norm condition number for the S matrix.
-func (f SVDFactors) Cond() float64 {
-	return f.Sigma[0] / f.Sigma[min(f.m, f.n)-1]
+// VFromSVD extracts the matrix V from the singular value decomposition, storing
+// the result in-place into the receiver. V is size n×n if svd.Kind() == SVDFull,
+// of size n×min(m,n) if svd.Kind() == SVDThin, and VFromSVD panics otherwise.
+func (m *Dense) VFromSVD(svd *SVD) {
+	r := svd.vt.Rows
```
Comment (kortschak) [inline mat64/svd.go]: Early panic for kind == 0.

File: mat64/svd.go (hunk 2 above)
Comment (btracey) [inline mat64/svd.go]: If I add "for the SVDKind values" is that sufficient?

Comment (kortschak) [inline mat64/svd.go]: Perfectly.

File: mat64/svd.go (hunk 1 above)
Comment (btracey) [inline mat64/svd.go]: I don't know. I'm not volunteering :)

Comment (kortschak) [inline mat64/svd.go]: Sorry, I should be more clear. Do you see any utility for a cmat128 SVD? If there is, then these should go in matrix. If there isn't then maybe they should go in matrix.

Comment (btracey) [inline mat64/svd.go]: Done.

File: mat64/svd.go (hunk 2 above)
Comment (btracey) [inline mat64/svd.go]: Done.

File: mat64/svd.go (hunk 3 above)
Comment (btracey) [inline mat64/svd.go]: Done.

File: mat64/svd.go (hunk 4 above)
Comment (btracey) [inline mat64/svd.go]: Done.

File: mat64/svd.go (hunk 5 above)
Comment (btracey) [inline mat64/svd.go]: Done.

File: mat64/svd.go (hunk 6 above)
Comment (btracey) [inline mat64/svd.go]: Done.

**Comment by btracey** [comment]
PTAL

File: consts.go [hunk 7]
```diff
@@ -0,0 +1,24 @@
+// Copyright ©2015 The gonum Authors. All rights reserved.
```
Comment (kortschak) [inline consts.go]: 2016?

File: mat64/svd.go [hunk 8]
```diff
@@ -6,474 +6,165 @@
...
-					v.set(i, k, e[i])
-				}
-			}
+		svd.vt = blas64.General{
+			Rows:   min(m, n),
+			Cols:   n,
+			Stride: n,
+			Data:   make([]float64, min(m, n)*n),
 		}
+		jobU = lapack.SVDInPlace
```
Comment (kortschak) [inline mat64/svd.go]: Will these change when/if the lapack implementation is extended?

File: consts.go (hunk 7 above)
Comment (btracey) [inline consts.go]: Oops. I changed it somewhere and it's not one of the files in this PR, so I wonder which of my files I changed...

File: mat64/svd_test.go [hunk 9]
```diff
@@ -5,207 +5,173 @@
...
+		}
+		m, n := test.a.Dims()
+		sigma := NewDense(min(m, n), min(m, n), nil)
+		for i := 0; i < min(m, n); i++ {
+			sigma.Set(i, i, s[i])
+		}
 
-			sigma: []float64{21.259500881097434, 1.5415021616856566, 1.2873979074613628},
+		var ans Dense
+		ans.Product(u, sigma, v.T())
```
Comment (kortschak) [inline mat64/svd_test.go]: Nice.

File: mat64/svd.go (hunk 8 above)
Comment (btracey) [inline mat64/svd.go]: Yes, they probably should. It's more memory efficient to have one of them be written in place into A (the larger one). TODO added.

**Comment by kortschak** [comment]
LGTM but consider waiting for @vladimir-ch.

File: mat64/svd_test.go (hunk 9 above)
Comment (btracey) [inline mat64/svd_test.go]: Indeed! Thanks for coding that!

File: mat64/svd.go [hunk 10]
```diff
@@ -6,474 +6,171 @@
...
+// where Σ is an m×n diagonal matrix of singular vectors, U is an m×m unitary
+// matrix of left singular vectors, and V is an n×n matrix of right singular vectors.
 //
-//  sigma[0] >= sigma[1] >= ... >= sigma[n-1].
+// It is frequently not necessary to compute the full SVD. Computation time and
+// storage costs can be reduced using the appropriate kind. Only the singular
+// values can be computed (kind == SVDNone), or a "thin" representation of the
+// singular values (kind = SVDThin). The thin representation can save a significant
+// amount of memory if m >> n. See the documentation for the lapack.SVDKind values
+// more information.
```
Comment (vladimir-ch) [inline mat64/svd.go]: **for** more information ?

Comment (btracey) [inline mat64/svd.go]: Done.

File: mat64/svd.go [hunk 11]
```diff
@@ -6,474 +6,171 @@
...
+func (svd *SVD) Cond() float64 {
+	return svd.s[0] / svd.s[len(svd.s)-1]
+}
 
-			m: m, n: n,
-		}
+// Values returns the singular values of the factorized matrix in decreasing order.
+// If the input slice is non-nil, the values will be stored in-place into the slice.
+// In this case, the slice must have length min(m,n), and Values will panic with
+// matrix.ErrSliceLengthMismatch if it is not. If the input slice is nil,
```
Comment (vladimir-ch) [inline mat64/svd.go]: s/if it is not/otherwise/ ?

File: mat64/svd.go [hunk 12]
```diff
@@ -6,474 +6,171 @@
...
+	return svd.kind
+}
 
-	if trans {
-		return SVDFactors{
-			U:     v,
-			Sigma: sigma,
-			V:     u,
+// Cond returns the 2-norm condition number for the factorized matrix. Cond will
+// panic if the receiver does not contain a successful factorization.
```
Comment (vladimir-ch) [inline mat64/svd.go]: `Cond` does not seem to panic.

File: mat64/matrix.go [hunk 13]
```diff
@@ -295,11 +295,8 @@ func Cond(a Matrix, norm float64) float64 {
 	case 1:
 		lnorm = lapack.MaxColumnSum
 	case 2:
-		// Use SVD to compute the condition number.
-		// TODO(btracey): Replace this with temporary workspace when SVD is fixed.
-		tmp := NewDense(m, n, nil)
-		tmp.Copy(a)
-		svd := SVD(tmp, math.Pow(2, -52.0), math.Pow(2, -966.0), false, false)
+		var svd SVD
+		svd.Factorize(a, matrix.SVDNone)
```
Comment (vladimir-ch) [inline mat64/matrix.go]: What if the factorization fails? It can fail, can't it? Then `Cond` would panic.

Comment (btracey) [inline mat64/matrix.go]: I'm having a tough time seeing if the decomposition only fails when the input matrix is singular or near-singular. I want to say that if it fails, we can return math.Inf(1).

Comment (vladimir-ch) [inline mat64/matrix.go]: Returning inf sounds right to me.

Comment (btracey) [inline mat64/matrix.go]: Done.

File: mat64/svd.go (hunk 12 above)
Comment (btracey) [inline mat64/svd.go]: Done.

File: mat64/svd.go (hunk 11 above)
Comment (btracey) [inline mat64/svd.go]: Done.

**Comment by btracey** [comment]
PTAL

**Comment by vladimir-ch** [comment]
So nice. LGTM

**Comment by kortschak** [comment]
Yes, nice work. We had decided to leave ED for later, no? There are minor things to either commit or discuss wrt API stability now.

**Comment by btracey** [comment]
We had decided to transform EigenFactors into Eigen, with a signature as it would be with Lapack, just without an actual Lapack implementation backing it. I can do that, it seems reasonably straightforward. The only issue is I seem to remember there was a problem with Eigen returning bad values. I can't find the discussion of that though.

Do you have a list of API stability issues remaining? I think we've accomplished all of the ones that were on my list. If you do have a list, maybe open a "Reach 1.0" issue with the outstanding problems. I'll take another read of the proposed documents.

**Comment by kortschak** [comment]
> We had decided to transform EigenFactors into Eigen, with a signature as it would be with Lapack, just without an actual Lapack implementation backing it. I can do that, it seems reasonably straightforward. The only issue is I seem to remember there was a problem with Eigen returning bad values. I can't find the discussion of that though.

That sounds familiar.

> Do you have a list of API stability issues remaining? I think we've accomplished all of the ones that were on my list. If you do have a list, maybe open a "Reach 1.0" issue with the outstanding problems. I'll take another read of the proposed documents.

https://github.com/gonum/matrix/milestones/API%20stability

**Comment by btracey** [comment]
134 is not in the way of stability, is it? It's only adding to the API.

**Comment by kortschak** [comment]
A decision needs to be made as to whether cmat128.Matrix has an H method or a T method that == H. That impacts on details of matrix.

**Comment by btracey** [comment]
Right, which is 283.

**Comment by kortschak** [comment]
Yes, but 238 blocks 134.

**Comment by btracey** [comment]
Woo!

File: mat64/svd.go [hunk 14]
```diff
@@ -6,474 +6,174 @@
...
 }
 
-// SVD performs singular value decomposition for an m-by-n matrix a. The
-// singular value decomposition is an m-by-n orthogonal matrix u, an n-by-n
-// diagonal matrix s, and an n-by-n orthogonal matrix v so that a = u*s*v'. If
-// a is a wide matrix a copy of its transpose is allocated, otherwise a is
-// overwritten during the decomposition. Matrices u and v are only created when
-// wantu and wantv are true respectively.
+// Factorize computes the singular value decomposition (SVD) of the input matrix
+// A. The singular values of A are computed in all cases, while the singular
```
Comment (kortschak) [inline mat64/svd.go]: Ooops. Should probably say that A is an m×n.

**Pull request merged**